### PR TITLE
8264760: JVM crashes when two threads encounter the same resolution error

### PR DESCRIPTION
**Pull request by Wanghuang-Huawei**
As shown in JDK-8264760, I changed notes with @dholmes-ora and only fixed this issue by deleting the assert. ~~The other whole bigger issue will be fixed in the other issue.~~

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264760](https://bugs.openjdk.java.net/browse/JDK-8264760): JVM crashes when two threads encounter the same resolution error

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Harold Seigel](https://openjdk.java.net/census#hseigel) (@hseigel - **Reviewer**)

### Contributors
 * Wang Huang `<whuang@openjdk.org>`
 * Wu Yan `<wuyan34@huawei.com>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3392/head:pull/3392` \
`$ git checkout pull/3392`

Update a local copy of the PR: \
`$ git checkout pull/3392` \
`$ git pull https://git.openjdk.java.net/jdk pull/3392/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3392`

View PR using the GUI difftool: \
`$ git pr show -t 3392`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3392.diff">https://git.openjdk.java.net/jdk/pull/3392.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:58]: The use of CDL doesn't guarantee that both threads have reached the await() before the main thread does the countDown(). If you use a CyclicBarrier instead it will trip when the second thread arrives.

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:62]: Please add a comment:

// Try to have all threads trigger the nesthost check at the same time

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:64]: Please follow this by:
`throw new RuntimeException("Did not get expected IllegalAccessError");`

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Wang,

Thanks for reporting this and offering to fix it. I have a comment on the actual fix and a few issues with the testcase - though thanks for adding one. With a race condition like this it is hard to write a test that will fail reliably without the fix.

Thanks,
David

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:51]: "throws Throwable" is only necessary if you want unlisted checked exceptions to propagate.

File: src/hotspot/share/classfile/systemDictionary.cpp
Comment (dholmes-ora) [reply src/hotspot/share/classfile/systemDictionary.cpp:1939]: Sorry the above is not correct. I believe the code you have is correct. I will double-check some details then update my review.

Comment (dholmes-ora) [inline src/hotspot/share/classfile/systemDictionary.cpp:1939]: Please add a comment before line 1939:
// An existing entry means we had a true resolution failure (LinkageError) with our nest host, but we
// still want to add the error message for the higher-level access checks to report. We should
// only reach here under the same error condition, so we can ignore the potential race with setting
// the message. If we see it is already set then we can ignore it.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
I think the code you have is correct. I've suggested some additional commentary. There are a couple of races possible here, but I think the remaining ones are benign and we do not need to try and impose any stricter synchronization.

Thanks,
David

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:58]: I have tested `CyclicBarrier ` here and found that `CountDownLatch` could trigger this bug fast while `CyclicBarrier ` couldn't always trigger this bug.

Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:58]: @dholmes-ora

Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:58]: > The use of CDL doesn't guarantee that both threads have reached the await() before the main thread does the countDown(). If you use a CyclicBarrier instead it will trip when the second thread arrives.

I have tested a small case  the `CyclicBarrier ` version and `CountDownLatch ` version on `JDK17` & `JDK11`. It seems that :
* For `JDK17`, the wake-up time between two threads in the `CyclicBarrier ` version is bigger than the  `CountDownLatch ` version. In other words,   the  `CountDownLatch ` version can trigger this bug more easily .
* the `CyclicBarrier ` version and   the  `CountDownLatch ` version running under `JDK11`  have the same performance.
```java
// CountDownLatch version 
public class TestCountDownLatch {

  public static void main(String args[]) {
    CountDownLatch latch1 = new CountDownLatch(1);
    CountDownLatch latch2 = new CountDownLatch(2);

    MyThread t1 = new MyThread(latch1, latch2);
    MyThread t2 = new MyThread(latch1, latch2);

    t1.start();
    t2.start();

    try {
      // waiting thread creation
      latch2.await();
      latch1.countDown();

      t1.join();
      t2.join();
    } catch (InterruptedException e) {}   

    System.out.println(Math.abs(t1.getAwakeTime() - t2.getAwakeTime()));
  }

  static class MyThread extends Thread {
    private CountDownLatch latch1;
    private CountDownLatch latch2;
		private long awaketime;

    MyThread(CountDownLatch latch1, CountDownLatch latch2) {
      this.latch1 = latch1;
      this.latch2 = latch2;
    }

    @Override
    public void run() {
      try {
        latch2.countDown();
        // Try to have all threads trigger the nesthost check at the same time
        latch1.await();
        awaketime = System.nanoTime();
      } catch (InterruptedException e) {}
    }

    public long getAwakeTime() {
      return awaketime; 
    }
  }
}
```

```java
// CyclicBarrier version 
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.CyclicBarrier;

public class TestCyclicBarrier {

  public static void main(String args[]) {
    CyclicBarrier barrier = new CyclicBarrier(2);

    MyThread t1 = new MyThread(barrier);
    MyThread t2 = new MyThread(barrier);

    t1.start();
    t2.start();

    try {
      t1.join();
      t2.join();
    } catch (InterruptedException e) {}   

    System.out.println(Math.abs(t1.getAwakeTime() - t2.getAwakeTime()));
  }

  static class MyThread extends Thread {
    private CyclicBarrier barrier; 
		private long awaketime;

    MyThread(CyclicBarrier barrier) {
      this.barrier = barrier;
    }

    @Override
    public void run() {
      try {
        // Try to have all threads trigger the nesthost check at the same time
        barrier.await();
        awaketime = System.nanoTime();
      } catch (Exception e) {}
    }

    public long getAwakeTime() {
      return awaketime; 
    }
  }
}
```
* results:  
 - `CountDownLatch` against `CyclicBarrier`
 ```
  CountDownLatch: 3960
  CyclicBarrier: 3361280
  CountDownLatch: 6120
  CyclicBarrier: 3337540
  CountDownLatch: 6160
  CyclicBarrier: 3462920
  CountDownLatch: 9150
  CyclicBarrier: 3328090
  CountDownLatch: 6580
  CyclicBarrier: 3345450
  CountDownLatch: 6340
  CyclicBarrier: 3342900
  CountDownLatch: 1330
  CyclicBarrier: 3379210
  CountDownLatch: 7780
  CyclicBarrier: 3219020
  CountDownLatch: 2460
  CyclicBarrier: 3297020
  CountDownLatch: 7320
  CyclicBarrier: 3332770
```
-  `JDK17` against `CyclicBarrier`
```
CyclicBarrier jdk17: 3188590
CyclicBarrier jdk11: 49090
CyclicBarrier jdk17: 3123340
CyclicBarrier jdk11: 14680
CyclicBarrier jdk17: 3107910
CyclicBarrier jdk11: 780
CyclicBarrier jdk17: 3072600
CyclicBarrier jdk11: 1720
CyclicBarrier jdk17: 3164340
CyclicBarrier jdk11: 41020
CyclicBarrier jdk17: 3098490
CyclicBarrier jdk11: 7060
CyclicBarrier jdk17: 3058220
CyclicBarrier jdk11: 14750
CyclicBarrier jdk17: 3052460
CyclicBarrier jdk11: 660
CyclicBarrier jdk17: 3083650
CyclicBarrier jdk11: 14670
CyclicBarrier jdk17: 3116260
CyclicBarrier jdk11: 850
```

File: test/hotspot/jtreg/runtime/Nestmates/membership/HostNoNestMember.jcod
Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/HostNoNestMember.jcod:1]: OK. I will fix that.

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:51]: Thank you for your review. `throws Throwable` will be removed in my next commit.

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:38]: I split this class into a single file `HostNoNestMember.java` as @hseigel suggested and add a comment.

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:53]: Thank you. It will be deleted in next commit.

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:43]: Sure. Thank you.

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:68]: Thank you for your review. It will be more clear after extracting the logic. I will refact that.

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:62]: Thank you. I will add that in next commit.

File: src/hotspot/share/classfile/systemDictionary.cpp
Comment (Wanghuang-Huawei) [reply src/hotspot/share/classfile/systemDictionary.cpp:1939]: OK. Thank you for your advice. I will add that.

**Comment by Wanghuang-Huawei** [comment]
Thank you for your review.
> Class HostNoNestMember could be removed entirely from class TestNestHostErrorWithMultiThread by doing the following:
> 
> Generate a jcod representation for class HostNoNestMember$Member and add it to HostNoNestMember.jcod.
I think a java file is more clear?
> 
> Delete class HostNoNestMember from TestNestHostErrorWithMultiThread.java.
> 
 I split `HostNoNestMember.java` from `TestNestHostErrorWithMultiThread.java`.
> Change the test to compile HostNoNestMember.jcod before compiling TestNestHostErrorWithMultiThread.java.
> 
> Doing this removes the confusion of class HostNoNestMember being compiled twice.

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:45]: Nit: you only need a single Runnable instance as it is stateless. The Runnable could even be defined inline rather than as a separate named class.

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:51]: Catch block should have:

throw new Error("Unexpected interrupt");

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:79]: Catch block should have:

throw new Error("Unexpected interrupt");

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:70]: After the call to test you should have:

throw new Error("IllegalAccessError was not thrown as expected");

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:29]: You don't use the test library

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:33]: No need for othervm in this test.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi,

Sorry for the delayed re-review. There are still a bunch of nits with the test logic that I'd like to see addressed. Sorry I should have flagged them earlier.

Thanks,
David

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:42]: Sure. It will make the code more clear. I will do that.Thank you.

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:45]: It's my fault. I will adjust that. Thank you.

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:33]: Sure. I will do that.

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:29]: Yes. Thank you for your review.

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:70]: Thank you. I will change that.

**Review by dholmes-ora — APPROVED** [review]
Looks good. Thanks for your patience here.

David

**Comment by dholmes-ora** [comment]
The new test is failing for a reason that is not clear to me yet.

**Comment by dholmes-ora** [comment]
I was unable to reproduce the test failure locally.

**Comment by Wanghuang-Huawei** [comment]
> The new test is failing for a reason that is not clear to me yet.

I will run regression and check this thing.

**Comment by Wanghuang-Huawei** [comment]
> I was unable to reproduce the test failure locally.

Thank you for your feedback. I have reproduced this problem (~ 10% ) and fixed it.  Could you give me a favor to review new commit ? Thank you very much.

**Review by dholmes-ora — APPROVED** [review]
I don't see exactly why the lack of join() would lead to the test failure, but doing the join() is cleaner and consistent with what other tests do.

Thanks,
David

**Comment by Wanghuang-Huawei** [comment]
> I don't see exactly why the lack of join() would lead to the test failure, but doing the join() is cleaner and consistent with what other tests do.
> 
> Thanks,
> David

The reason is that the main thread needs to exit and wait for the other two threads to exit. However, the `jtreg` just waits for a fix time. If two threads are not scheduled in this time, an error will be triggered.

**Comment by dholmes-ora** [comment]
We actually need a second review here (hotspot group rules).

Thanks,
David

**Comment by Wanghuang-Huawei** [comment]
> hotspot group rules

Sure. It's my fault. Anyway to revoke `/integrate` here?

**Comment by dholmes-ora** [comment]
No need to revoke it, the sponsor just needs to hold off until there is a second review.

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (hseigel) [inline test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:34]: Please add a comment here that HostNoNestMember.jcod must be compiled after HostNoNestMember.java because the class file from the jcod file must replace the HostNoNestMember class file generated from HostNoNestMember.java.

File: test/hotspot/jtreg/runtime/Nestmates/membership/HostNoNestMember.java
Comment (hseigel) [inline test/hotspot/jtreg/runtime/Nestmates/membership/HostNoNestMember.java:29]: Can you modify the above comment to something such as:

 * This class was used to produce a jcod file in which the
 * NestMembers attribute was modified to make it empty. Class
 * HostNoNestMember$Member has class HostNoNestMember as its
 * NestHost, which will trigger an error when resolving.
 *
 * When compiled, this generates a HostNoNestMember class and
 * a HostNoNestMember$Menber class.  The former class, HostNoNestMember,
 * gets overwritten when HostNoNestMember.jcod gets compiled.

File: test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java
Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/TestNestHostErrorWithMultiThread.java:34]: Thank you for your review. I will add that.

File: test/hotspot/jtreg/runtime/Nestmates/membership/HostNoNestMember.java
Comment (Wanghuang-Huawei) [reply test/hotspot/jtreg/runtime/Nestmates/membership/HostNoNestMember.java:29]: OK. I will fix that. Thank you.

**Review by hseigel — APPROVED** [review]
Changes look good.  Thanks for fixing this.
Harold

**Review by dholmes-ora — APPROVED** [review]

**Comment by Wanghuang-Huawei** [comment]
> Changes look good. Thanks for fixing this.
> Harold

Thank you for your approval. @dholmes-ora @hseigel

**Pull request closed without merging**
